qiskit_experiments/database_service/db_experiment_data.py

```diff
@@ -1156,26 +1212,73 @@ def block_for_results(self, timeout: Optional[float] = None) -> "DbExperimentDat
         Returns:
             The experiment data with finished jobs and post-processing.
         """
-        if self._analysis_futures:
-            self._wait_for_analysis(timeout)
-        else:
-            self._wait_for_jobs(timeout)
-        self._removed_done_futures()
+        start_time = time.time()
+        with self._job_futures.lock and self._analysis_futures.lock:
+            # Lock threads to get all current job and analysis futures
+            # at the time of function call and then release the lock
+            job_ids = self._job_futures.keys()
+            job_futs = self._job_futures.values()
+            analysis_ids = self._analysis_futures.keys()
+            analysis_futs = self._analysis_futures.values()
+
+        # Wait for futures
+        self._wait_for_futures(job_futs + analysis_futs, name="jobs and analysis", timeout=timeout)
+
+        # Clean up done job futures
+        num_jobs = len(job_ids)
+        for jid, fut in zip(job_ids, job_futs):
+            if (fut.done() and not fut.exception()) or fut.cancelled():
+                if jid in self._job_futures:
+                    del self._job_futures[jid]
+                    num_jobs -= 1
+
+        # Clean up done analysis futures
+        num_analysis = len(analysis_ids)
+        for cid, fut in zip(analysis_ids, analysis_futs):
+            if (fut.done() and not fut.exception()) or fut.cancelled():
+                if cid in self._analysis_futures:
+                    del self._analysis_futures[cid]
+                    num_analysis -= 1
+
+        # Check if more futures got added while this function was running
+        # and block recursively. This could happen if an analysis callback
+        # spawns another callback or creates more jobs
+        if len(self._job_futures) > num_jobs or len(self._analysis_futures) > num_analysis:
```

Comment:
Because futures terminated with exception still live, likely this evaluation can be true without adding more jobs. Then this doesn't cause dead lock?